### ISSUE TITLE
[Feature Request] Crop element select

### ISSUE THREAD

**Issue by Mystqr4** [WindowTop/WindowTop-App#324]
I know its a long shot and probably something more down the line if at all but wonder if it is at all possible to add an element select when attempting to crop. It would come in so handy when trying to accurately crop a particularly frame out for example a video window where it could automatically find the video frame on hover and be able to discern the difference between the video frame and the black bars too but also be able to auto detect the video frame with the bars too for those times and people that need it for subtitles.

I guess in similar fashion to the way firefox's default screenshot tool works where it can highlight a certain frame as it finds it. I realise the construct of a website with CSS and frames make it far easier for that to be a thing though but I have no idea what is possible and what isn't but figured it would very useful and make everything much more efficient and accurate at times.

I'm not sure of it could be useful for some but I feel it would just add as an enhancement to what is currently available and make everything a little more fluid without taking away the custom hand selected crop feature either. It seems like it would be a pretty complex thing to add though so I am not sure. I thought it was maybe worth mentioning as something to look at perhaps in the future.

**Comment by gileli121** [comment]
Good idea, and yes it is possible.
The part of detecting the objects is not as complex as you think.
I know how to do it. In fact, this is what I already did for implementing some existing 
Functionality.

I have seen your request done using AI, but this is because some developers are not too familiar with Windows API.

The complex part here is about implementing the UI for it.

**Comment by Mystqr4** [comment]
> Good idea, and yes it is possible. The part of detecting the objects is not as complex as you think. I know how to do it. In fact, this is what I already did for implementing some existing functionality.
> 
> I have seen your request done using AI, but this is because some developers are not too familiar with Windows API.
> 
> The complex part here is about implementing the UI for it.

Extraordinary! This would be an excellent achievement if it works out the way I suspect it would. It seems like you are already knew exactly what I was thinking without the entire explanation on my part.

The fact that you can do it it without AI makes it even better to me too.

I'm not sure how difficult it will be to add the UI for it but I am sure you will do a great job.

**Comment by gileli121** [comment]
@Mystqr4 

> It seems like you are already knew exactly what I was thinking without the entire explanation on my part.

I still did not thought about it in detail yet. Bu currently, I see two approaches - the first one is the "snapping" approach, and the other one is the select-element approach.

## Snapping approach
In this approach, I suggest that while you drag the crop selector, it will snap to the edge of the elements. It may be easier to develop.

## Select-element approach
In this approach, you basically select an element similar to how you select a dom element in Google Chrome developer tools, and youcrop to it..

**Comment by Mystqr4** [comment]
Both are good options. 

Snapping approach:
The crop selector whilst easier to implement might annoy people if it constantly snaps to the edge of the wrong element and may lead to frustrations.

Select-Element approach:
This one will probably be best if you can modify the the element select by dragging key points like you can with the firefox screenshot tool. Most times people won't have to but I imagine being able to would help signficantly.

Whatever way you go I feel like it would be a huge improvement and not just for people with made eye sight or less steady hands.

I honestly hope that sales do go up regardless of this potentially new enhancement . The amount of development effort you put into the current release was phenomenal. 
Buying this software has been well worth the cost. It took me a while to come around to buying it but I am so glad I did. Last I remember it was well priced and is a lifetime license which entitles you to use it on 3 computers. Can't argue with that really.

**Comment by gileli121** [comment]
Probably should switch to this 
https://sales.jetbrains.com/hc/en-gb/articles/207240845-What-is-a-perpetual-fallback-license-

(without the subscription part)
Anyway, any user who purchased when it was "lifetime" will keep enjoying it.


```
Select-Element approach:
This one will probably be best if you can modify the the element select by dragging key points like you can with the firefox screenshot tool. Most times people won't have to, but I imagine being able to would help signficantly.
```

OK, makes sense. I hope to work on it in the next few months

**Comment by vertigo220** [comment]
Not sure if helpful, but maybe take a look at ShareX for code and implementation ideas. It does this very well.

**Comment by gileli121** [comment]
@vertigo220 

Thanks for the information.  
I just tested it, and indeed, the **Capture -> Region** option does detect regions within a window.  

However, the detection is quite basic—for example, when playing a **YouTube** video in **Edge**,  
their selector doesn't isolate the video player but rather detects the entire browser window, excluding the **tabs**.  

From my knowledge and experience, I know they didn't push this to its full potential.  
So, I don't see the need to take inspiration from their implementation—they didn't do a good enough job.


Regarding this **feature request**, I have good news:  
For several days now, I’ve been thinking that it deserves **top priority** in the **roadmap** (#389).  

And now, after seeing more people responding here, I’ve decided to prioritize it second, right after #183.  
I’ve moved this **feature request** to follow #183 under the **Big Features** list in the roadmap.  

The reason #183 remains first is that it should be easy to implement and will provide a great addition to the toolbar. However, since it should be a quick task, I don’t think it will take long before I get to this one.

**Comment by gileli121** [comment]
@Mystqr4 @vertigo220 

I’m happy to share that I’ve started working on this feature over the past few days.
I’m currently in the very early stages of a proof of concept for the element detection mechanism itself (which, as it turns out, already partially exists in code used to manage the mouse cursor icon during mouse interaction mode).

In any case, this proof-of-concept phase looks very promising. Here’s a recording of the first POC:

![Image](https://github.com/user-attachments/assets/92935123-99c5-45aa-9e77-9d7db7dd2ae8)

Here’s an example of a more practical case — detecting a video player, for instance (which is more likely in cases where cropping is used):

![Image](https://github.com/user-attachments/assets/8b4418b2-9b83-4c72-af9e-6113a4a32c88)

From this, it’s clear that developing this requested feature is definitely possible :)

**Comment by Mystqr4** [comment]
This looks both very encouraging and impressive at the same time. I would love to see if it could further distinguish viewable video from black bars and capture the variables of with and without the black bars on some movies/videos depending on where you hover the mouse over the video.
Nonetheless its amazing progress already.

It's good to hear that some of the code already exists which should definitely help.

**Comment by gileli121** [comment]
@Mystqr4, Today I continue to work on it. here is progress from today

https://github.com/user-attachments/assets/851583d6-a972-4cfc-86df-8f990959a523

Let me know what you think and how it looks :)

**Comment by gileli121** [comment]
Another demo:

https://github.com/user-attachments/assets/3a61d307-8176-4932-808d-506244a413b6

**Comment by gileli121** [comment]
@Mystqr4 
Progress from today. this video show much more the capabilities :)

https://github.com/user-attachments/assets/0dda9567-e263-490e-8e25-765b7e80b955

**Comment by Mystqr4** [comment]
Forgive me for I have been having troubles with my computer but so far it looks very impressive and you seem to be moving swiftly through this.

I like that you have also retained the option to drag the cursor to specify one's own crop if one still desires.

Honestly now that I see your proof of concepts and initial testing I can see a multitude of use cases. I can see that this could be extremely helpful for in productivity situations where one could essential PiP elements from GUI's from programs like gimp for example and then set it on top and interactive mode to allow one have frequently used elements where they want it on screen. This could be a game changer for many.

This is very impressive. I know interactive mode can sometimes be an issue for some buttons but it still could be very useful. There are so many use cases for this.

This is excellent work. I am looking forward to seeing more of this. You are absolutely exceeding expectations.

**Comment by gileli121** [comment]
@Mystqr4 I’m really happy to hear this feedback :)
Indeed, it took me several days just to think about how to make it as clear as possible for users without interfering with other functionality. At first, I couldn’t figure out how to do it without adding yet another button along the way. The UI approach I came up with seemed very clear and met the requirements of being simple, intuitive, and not at the expense of what already exists.
I’m glad I’m exceeding your expectations — though it’s not really surprising :) I’m very experienced at what I do, and there’s simply no competition to the level I bring with WindowTop.
If I didn’t have enough users and purchases coming in at a steady pace, I would’ve abandoned this project long ago. Users like you, and those who support the product, are the main reason I keep working on it. So thank you.

**Comment by gileli121** [comment]
I’m happy to announce that I’ve released a new version. It’s currently marked as a BETA version until it’s proven stable. You’re welcome to try it:

https://github.com/WindowTop/WindowTop-App/releases/tag/5.29

Let me know how its works :)

**Comment by gileli121** [comment]
@vertigo220 @Mystqr4 

I’m happy to announce that I’ve released a new version. It’s currently marked as a BETA version until it’s proven stable. You’re welcome to try it:

https://github.com/WindowTop/WindowTop-App/releases/tag/5.29

Let me know how its works :)

**Comment by Mystqr4** [comment]
So far in testing I have noticed that the following:

-MPV:
Crop does not pick up the video screen as an element however it does pick up the menu bar items as an element.

- VLC: (Latest version)
Crop works on the video screen as an element however when I PiP it and then use the F command to fullscreen the video within the PiP it becomes wildly off center and enlarged in a random spot if that makes sense.

- QMplay2 (Latest version)
I have at least once seen the PiP from element select turn into a square with what seems like an extremely zoomed in image of the video with pixels. Upon fullscreening with F it exhibited the same issue as VLC.

- MPC-HC (Latest version by CLSID)
Same problems as VLC on Fullscreen with off-centered fullscreen. Fullscreen is not affected when just PiP without crop and works as expected.
I have noticed that when cropping and PiP it will find the element crop and then PiP but not PiP to a smaller size. It's basically the opposite of regular PiP where I will usually have to decide if I should enlargen the PiP window but with PiP crop it essentially stays the same size it was before but cropped so I will have to shrink down the size of the video to be smaller and in the corner of my screen. I am not sure if this is working as intended but it is something I noticed and thought I would mention it.
I have also noticed that the tab on the top of the Window with the title does not appear on MPC-HC when Fullscreened within a PiP.

- MPC-BE (Latest version)
Cannot detect the video screen element. captures menu bar buttons if needed but not anything else.

- MPC-QT (Latest version)
Cannot detect the video screen element. captures menu bar buttons if needed but not anything else.

- SMplayer (Latest version)
Captures elements including video. Fullscreen issue happens as with other players. I have also noticed with this player that even if I am not in PiP Intereact mode I can still use the mouse to pause and play and double clicking on the video whilst in PiP will cancel the PiP rather than bring up a fullscreen. To attempt to fullscreen in PiP I have to use the F key which is pretty common to fullscreen many of these video players.

- In General Use: 
I have noticed that WIndowTop now seems to get confused after making a crop PiP and a regular PiP. By that I mean if I attempt a crop PiP and then try it out and then cancel it only to try a regular PiP it will no go to the regular smaller PiP.
If I close the program and try a regular PiP I will it will work as expected and be a small PiP as per normal but once again if I follow the same path as before and attempt a cropped PiP and then a regular PiP the problem will occur again.

I am imaginig some of the problems with these video players comes down to output renderer for example with MPC-BE I am using MPC Video Renderer but in MPC-HC I am using a different output renderer (MadVR).

**Comment by gileli121** [comment]
@Mystqr4 
Thanks for testing. sorry for the late reply. I was very busy during last 2 weeks.


Most of the issues you’re describing are related to video players. I don’t expect this to work 100% on video players, especially because of how they operate. It should work quite well on other programs like browsers.

Regarding this case:

> Crop works on the video screen as an element however when I PiP it and then use the F command to fullscreen the video within the PiP it becomes wildly off center and enlarged in a random spot if that makes sense.

Are you expecting that after you switch the video to fullscreen, the crop will automatically adjust itself to the new position on the element? If so, that sounds like an interesting feature. But that’s not how it works… once you’ve already activated the crop to some element and then the window changes a bit because fullscreen was enabled, in that situation there’s no logic that tries to update the crop to the new element position you cropped. so you will see offscreen.

Regarding this:

> In General Use:
I have noticed that WIndowTop now seems to get confused after making a crop PiP and a regular PiP. By that I mean if I attempt a crop PiP and then try it out and then cancel it only to try a regular PiP it will no go to the regular smaller PiP.
If I close the program and try a regular PiP I will it will work as expected and be a small PiP as per normal but once again if I follow the same path as before and attempt a cropped PiP and then a regular PiP the problem will occur again.

I couldn’t understand you. Is this something that didn’t happen in older versions? Could you record a short video showing the issue?

Thanks.

**Comment by gileli121** [comment]
@Mystqr4 
I released 5.29.1 with minor bugs.
Please try this one instead:
https://github.com/WindowTop/WindowTop-App/releases/tag/5.29

Thanks

**Comment by gileli121** [comment]
5.29.1 seems stable enough so I released it.
I know that its not perfect. but seems to work good in most cases

**Comment by gileli121** [comment]
Please open another issue for the other problems you reported and I hope to get them at some point
